_posts/2024-11-23-dotfiles.md

```diff
@@ -0,0 +1,275 @@
+your dotfiles. It will bring your machine to the desired end state -- your custom development
+environment. It takes no arguments. You just run it.
+
+```sh
+./install.sh
+Starting .dotfiles install
+```
+
+The first thing we're going to do is to edit the install script to follow some bash norms:
+
+```sh
+#!/usr/bin/env bash
+
+set -e # fail on error
+set -u # error on undefined variables
+
+echo "Starting .dotfiles install"
+```
+
+Without `-e` bash will happily keep going whenever an error happens. We want to [fail
+fast](https://en.wikipedia.org/wiki/Fail-fast_system) so that we don't continue when an command that
+subsequent ones depend on failed. As you hit failures you will need to program around them, and
+handle an error, depending on the context. This will make your dotfiles more relisient and will make
+you a better programmer.
+
+The `-u` option has a similar effect, failing the script if you try to dereference variables (e.g.
+`$MY_URL`) when the variable `MY_URL` does not exist.
+
+Commit your changes and push them. Years from now I'd suggest using something like
+[lazygit](https://github.com/jesseduffield/lazygit) to walk through your changes over time. It's
+good to see your progression in this way. `lazygit` is a wonderful tool. why don't we add it to your
+dotfiles so you have it forever?
+
+# maintain your dotfiles
+
+Let's make a real change and get `lazygit` into your dotfiles. Starting out, keep it simple. Just
+append to `install.sh`. You'll know when it's time to break things up, but doing so right now is
+counterproductive.
+
+I'm on a mac so I'll be using the homebrew package manager. Add the following to your `install.sh`:
+
+```sh
+# If brew does not exist on this system, install it.
+if ! command -v brew >/dev/null; then
+    echo "Installing homebrew"
+    # from: https://brew.sh/
+    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+fi
+```
+
+Instead of just running the command to install brew every time, we check to see if it's installed
+first. If you're not familiar with the `if` conditional, `command -v brew` outputs the path of the
+`brew` command, or errors out. But what about the `set -e` we added earlier? If you are testing for
+the success of the command, as we are, an error will not fail, but will result in a `true` for our
+if statement because of the `!` negation.
+
+Run `./install.sh`. If you didn't have homebrew, you will now. Run it again, and it should not
+install it a second time. Commit and push. This is the kind of granular micro commit strategy that I
+think really works well for a dotfiles.
+
+And now, once we have that, it's time to install `lazygit`. Add this to your `install.sh`:
+
+```sh
+# update homebrew
+brew update --quiet
+# a sorted list of packages to install
+brew install --quiet --formula \
+    lazygit \
+    ;
+```
+
+We use the `\` syntax to break up a command onto multiple lines. The goal here is to have each
```

Comment:
I think this onto/into one could be subjective, not entirely sure did some looking and I think into is preferred but you are stacking lines so 🤷